File: app/src/main/java/com/Fachhochschulebib/fhb/pruefungsplaner/CheckListAdapter.java
package com.Fachhochschulebib.fhb.pruefungsplaner;

//////////////////////////////
// CheckListAdapter
//
// autor:
// inhalt:  erstellt die Checkliste für die Wahl der Studiengänge
// zugriffsdatum: 01.10.20
//
//
//////////////////////////////

import android.content.Context;
import android.content.SharedPreferences;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.TextView;
import android.widget.Toast;

import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import static android.content.Context.MODE_PRIVATE;


public class CheckListAdapter extends RecyclerView.Adapter<CheckListAdapter.ViewHolder> {
    private List<String> coursesList;
    private List<Boolean> chosenList;
    String selectedCourse;
    private Context context;
    private StartClass globalVariable;

    // Provide a suitable constructor (depends on the kind of dataset)
    public CheckListAdapter(List<String> courses,
                     List<Boolean> chosen,
                            Context ApplicationContext) {
        coursesList = courses;
        chosenList = chosen;
        context = ApplicationContext;
    }

    // Create new views (invoked by the layout manager)
    @Override
    public CheckListAdapter.ViewHolder onCreateViewHolder(ViewGroup parent,
                                                   int viewType) {
        // create a new view
        LayoutInflater inflater = LayoutInflater.from(
                parent.getContext());
        View v =
                inflater.inflate(R.layout.checkliste, parent, false);
        ViewHolder vh = new ViewHolder(v);

        globalVariable = (StartClass) context;

        SharedPreferences sharedPreferencesSelectedCourse = context.
                getSharedPreferences("validation", MODE_PRIVATE);
        selectedCourse = sharedPreferencesSelectedCourse.
                getString("selectedStudiengang","0");

        return vh;
    }

    private boolean addFavorite(int position) {
        if(!coursesList.get(position).equals(selectedCourse) || globalVariable.isChangeFaculty()) {
            chosenList.set(position,
                    !chosenList.get(position));
        } else {
            Toast.makeText(context, context.getString(R.string.favorite_main_course),
                    Toast.LENGTH_SHORT).show();
        }

        return chosenList.get(position);
    }

    // Replace the contents of a view (invoked by the layout manager)
    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position) {
        // Initialisierung der Komponenten
        holder.nameCourse.setText(coursesList.get(position));
        holder.checkBoxCourse.setChecked(chosenList.get(position));

        holder.checkBoxCourse.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // Speichere die auswahl in der Liste
                holder.checkBoxCourse.setChecked(addFavorite(position));
            }
        });

        holder.nameCourse.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // Speichere die auswahl in der Liste
                holder.checkBoxCourse.setChecked(addFavorite(position));
            }
        });

    }

    //Item anzahl
    @Override
    public int getItemCount() {
        return coursesList.size();
    }

    @Override
    public int getItemViewType(int position) {
        return position;
    }

    // Provide a reference to the views for each data item
    // Complex data items may need more than one view per item, and
    // you provide access to all the views for a data item in a view holder
    public class ViewHolder extends RecyclerView.ViewHolder {
        // each data item is just a string in this case
        private TextView nameCourse;
        private CheckBox checkBoxCourse;

        private ViewHolder(View v) {
            super(v);
            nameCourse = (TextView) v.findViewById(R.id.courseName);
            checkBoxCourse = (CheckBox) v.findViewById(R.id.checkBox);
        }
    }
}

File: app/src/main/java/com/Fachhochschulebib/fhb/pruefungsplaner/CheckListAdapter.kt
package com.Fachhochschulebib.fhb.pruefungsplaner

import android.content.Context
import androidx.recyclerview.widget.RecyclerView
import com.Fachhochschulebib.fhb.pruefungsplaner.StartClass
import android.view.ViewGroup
import android.view.LayoutInflater
import com.Fachhochschulebib.fhb.pruefungsplaner.R
import android.content.SharedPreferences
import android.view.View
import android.widget.Toast
import android.widget.TextView
import android.widget.CheckBox

//////////////////////////////
// CheckListAdapter
//
// autor:
// inhalt:  erstellt die Checkliste für die Wahl der Studiengänge
// zugriffsdatum: 01.10.20
//
//
//////////////////////////////
/**
 * Adapter-Class for the Recyclerview in the AddCourseFragment-Class
 *
 * @author Alexander Lange (Email:alexander.lange@fh-bielefeld.de)
 * @since 1.5
 *
 * @see AddCourseFragment
 * @see RecyclerView.Adapter
 */
class CheckListAdapter     // Provide a suitable constructor (depends on the kind of dataset)
    (
    private val coursesList: List<String>,
    private val chosenList: MutableList<Boolean>,
    private val context: Context
) : RecyclerView.Adapter<CheckListAdapter.ViewHolder>() {
    var selectedCourse: String? = null
    private var globalVariable: StartClass? = null

    /**
     * Called when Recyclerview needs a new ViewHolder.
     *
     * @param parent The ViewGroup into which the new View will be added after it is bound to
     *               an adapter position.
     * @param viewType The view type of the new View.
     *
     * @return A new ViewHolder that holds a View of the given view type.
     *
     * @author Alexander Lange
     * @since 1.5
     *
     * @see RecyclerView
     * @see RecyclerView.Adapter
     * @see RecyclerView.Adapter.onCreateViewHolder
     */
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ViewHolder {
        // create a new view
        // Get LayoutInflater
        val inflater = LayoutInflater.from(
            parent.context
        )
        // Inflate view to add to the Recyclerview
        val v = inflater.inflate(R.layout.checkliste, parent, false)
        //Create ViewHolder
        val vh = ViewHolder(v)

        globalVariable = context as StartClass

        //Get selected Course
        val sharedPreferencesSelectedCourse =
            context.getSharedPreferences("validation", Context.MODE_PRIVATE)
        selectedCourse = sharedPreferencesSelectedCourse.getString("selectedCourse", "0")

        //Return ViewHolder
        return vh
    }


    /**
     * Displays the data at a specific position.
     *
     * @param[holder] The ViewHolder which should be updated to represent the contents of the
     *        item at the given position in the data set.
     * @param[position] The position of the item within the adapter's data set.
     *
     * @author Alexander Lange
     * @since 1.5
     *
     * @see RecyclerView
     * @see RecyclerView.Adapter
     * @see RecyclerView.Adapter.onBindViewHolder
     */
    // Replace the contents of a view (invoked by the layout manager)
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        // Initialisierung der Komponenten
        holder.nameCourse.text = coursesList[position]
        holder.checkBoxCourse.isChecked = chosenList[position]
        holder.checkBoxCourse.setOnClickListener { // Speichere die auswahl in der Liste
            holder.checkBoxCourse.isChecked = addFavorite(position)
        }
        holder.nameCourse.setOnClickListener { // Speichere die auswahl in der Liste
            holder.checkBoxCourse.isChecked = addFavorite(position)
        }
    }


    /**
     * Selects/Unselects a course as favorite.
     *
     * @param[position] The position of the course in the list.
     *
     * @return true if the course was set as favorite,false if the course is no longer a favorite.
     * @author Alexander Lange
     * @since 1.5
     */
    private fun addFavorite(position: Int): Boolean {
        if (coursesList[position] != selectedCourse || globalVariable!!.isChangeFaculty) {
            chosenList[position] = !chosenList[position]
        } else {
            Toast.makeText(
                context, context.getString(R.string.favorite_main_course),
                Toast.LENGTH_SHORT
            ).show()
        }
        return chosenList[position]
    }

    /**
     * Returns the size of the courseList.
     *
     * @return the size of the courselist
     *
     * @author Alexander Lange
     * @since 1.5
     * @see RecyclerView
     * @see RecyclerView.Adapter
     * @see RecyclerView.Adapter.getItemCount
     */
    //Item anzahl
    override fun getItemCount(): Int {
        return coursesList.size
    }

    /**
     * Return the view type of the item at <code>position</code> for the purposes
     * of view recycling.
     *
     * @param[position] position to query
     * @return integer value identifying the type of the view needed to represent the item at
     *                 <code>position</code>. Type codes need not be contiguous.
     * @author Alexander Lange
     * @since 1.5
     * @see RecyclerView
     * @see RecyclerView.Adapter
     * @see RecyclerView.Adapter.getItemViewType
     */
    override fun getItemViewType(position: Int): Int {
        return position
    }

    /**
     * Inner class to provide a reference to the views for each data item
     * Complex data items may need more than one view per item, and
     * you provide access to all the views for a data item in a view holder
     *
     * @author Alexander Lange
     * @since 1.5
     */
    inner class ViewHolder internal constructor(v: View) : RecyclerView.ViewHolder(v) {
        // each data item is just a string in this case
        val nameCourse: TextView
        val checkBoxCourse: CheckBox

        init {
            nameCourse = v.findViewById<View>(R.id.courseName) as TextView
            checkBoxCourse = v.findViewById<View>(R.id.checkBox) as CheckBox
        }
    }
}